src/CodeGen_OpenGLCompute_Dev.cpp

```diff
@@ -0,0 +1,382 @@
+    } else if (ends_with(name, ".__block_id_z")) {
+        return "gl_WorkGroupID.z";
+    } else if (ends_with(name, ".__block_id_w")) {
+        internal_error << "4-dimension loops with " << name << " are not supported\n";
+    }
+    internal_error << "simt_intrinsic called on bad variable name: " << name << "\n";
+    return "";
+}
+
+int thread_loop_workgroup_index(const string &name) {
+    string ids[] = {".__thread_id_x",
+                    ".__thread_id_y",
+                    ".__thread_id_z",
+                    ".__thread_id_w"};
+     for (size_t i = 0; i < sizeof(ids) / sizeof(string); i++) {
+        if (ends_with(name, ids[i])) { return i; }
+     }
+     return -1;
+}
+}
+
+void CodeGen_OpenGLCompute_Dev::CodeGen_OpenGLCompute_C::visit(const Div *op) {
+    visit_binop(op->type, op->a, op->b, "/");
+}
+
+void CodeGen_OpenGLCompute_Dev::CodeGen_OpenGLCompute_C::visit(const Mod *op) {
+    visit_binop(op->type, op->a, op->b, "%");
+}
+
+void CodeGen_OpenGLCompute_Dev::CodeGen_OpenGLCompute_C::visit(const Cast *op) {
+    Type value_type = op->value.type();
+    // If both types are represented by the same GLSL type, no explicit cast
+    // is necessary.
+    if (map_type(op->type) == map_type(value_type)) {
+        Expr value = op->value;
+        if (value_type.code == Type::Float) {
+            // float->int conversions may need explicit truncation if the
+            // integer types is embedded into floats.  (Note: overflows are
+            // considered undefined behavior, so we do nothing about values
+            // that are out of range of the target type.)
+            if (op->type.code == Type::UInt) {
+                value = simplify(floor(value));
+            } else if (op->type.code == Type::Int) {
+                value = simplify(trunc(value));
+            }
+        }
+        value.accept(this);
+        return;
+    } else {
+        Type target_type = map_type(op->type);
+        print_assignment(target_type, print_type(target_type) + "(" + print_expr(op->value) + ")");
+    }
+}
+
+void CodeGen_OpenGLCompute_Dev::CodeGen_OpenGLCompute_C::visit(const For *loop) {
+    if (is_gpu_var(loop->name)) {
+        internal_assert(loop->for_type == ForType::Parallel) << "kernel loop must be parallel\n";
+        internal_assert(is_zero(loop->min));
+
+        debug(4) << "loop extent is " << loop->extent << "\n";
+        //
+        //  Need to extract workgroup size.
+        //
+        int index = thread_loop_workgroup_index(loop->name);
+        if (index >= 0) {
+            const IntImm *int_limit = loop->extent.as<IntImm>();
+            user_assert(int_limit != NULL) << "For OpenGLCompute workgroup size must be an constant integer.\n";
+            int new_workgroup_size = int_limit->value;
+            user_assert(workgroup_size[index] == 0 || workgroup_size[index] == new_workgroup_size) <<
+                "OpenGLCompute requires all gpu kernels have same workgroup size, "
```

Comment:
I think this error message should print out the dimension the difference occurs in.